File: SenLib/Sen3/FileFixes/book05_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class book05_dat : FileMod {
		public string GetDescription() {
			return "Fix typos in Imperial Chronicle.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xb2b35b9d531658a4ul, 0x5ed2a53477757d5aul, 0x72c066abu));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();
			var book = new BookTable(bin, EndianUtils.Endianness.LittleEndian);

			// Provincal -> Provincial
			book.Entries[3].Text = book.Entries[3].Text.Insert(514, "i");

			var bin2 = book.WriteToStream(EndianUtils.Endianness.LittleEndian);
			return new FileModResult[] { new FileModResult("data/scripts/book/dat_en/book05.dat", bin2) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xb2b35b9d531658a4ul, 0x5ed2a53477757d5aul, 0x72c066abu));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/book/dat_en/book05.dat", file) };
		}
	}
}

File: native/sen3/FileFixes/book05_dat.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "file_getter.h"

#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/book_table.h"
#include "sha1.h"
#include "util/stream.h"

namespace SenLib::Sen3::FileFixes::book05_dat {
std::string_view GetDescription() {
    return "Fix typos in Imperial Chronicle.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/book/dat_en/book05.dat",
            40305,
            SenPatcher::SHA1FromHexString("b2b35b9d531658a45ed2a53477757d5a72c066ab"));
        if (!file) {
            return false;
        }

        DuplicatableByteArrayStream bin(file->Data.data(), file->Data.size());
        BookTable book(bin, HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        // Provincal -> Provincial
        auto& entry3 = book.Entries.at(3).Text.value();
        entry3.insert(entry3.begin() + 514, 'i');

        std::vector<char> bin2;
        {
            MemoryStream ms(bin2);
            book.WriteToStream(ms, HyoutaUtils::EndianUtils::Endianness::LittleEndian);
        }

        result.emplace_back(SenPatcher::P3APackFile{
            std::move(bin2), file->Filename, SenPatcher::P3ACompressionType::LZ4});

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::book05_dat
